subprojects/docs/src/docs/userguide/authoringMaintainableBuildScripts.adoc

```diff
@@ -22,8 +22,62 @@ Gradle build scripts combine the qualities of declarative build logic, expressiv
 The third-party link:https://github.com/nebula-plugins/gradle-lint-plugin[Gradle lint plugin] helps with enforcing a desired code style in a build script if you are looking for appropriate linting automation.
 ====
 
-[[sec:declare_task_group_and_description]]
-=== Declare the group and description for tasks
+=== Avoiding Gradle internal APIs
+
+Use of Gradle internal APIs in plugins and build scripts causes builds to break when either Gradle or plugins change.
+
+The following packages are listed in the link:https://github.com/gradle/gradle/blob/180b9d3fa84b91768364c603380e82947437eda1/buildSrc/subprojects/configuration/src/main/kotlin/org/gradle/gradlebuild/public-api.kt[Gradle public API definition], with the exception of any subpackage with `internal` in the name:
```

Comment:
I fear that this list will become quickly outdated. Maybe we just use the link instead of listing the packages. Is there reason why you use specific commit hash in the URL?

Reply:
I use a specific commit hash because we might move that kt file and break the link. 

I would say that the list will become no smaller as we will not remove any packages from the public API. I think it's important that this list be included inline because it belongs in our documentation. Honestly, it should be more formal that this listing here, but I'm not sure it's worth that effort right now.

Reply:
Don't want to open another can of worms but could we inject the current commit hash and use that to build URLs? Or have special notation to resolve that automatically when processing the docs?

Thanks for clarifying.

Reply:
If we did, we would still not guarantee the link works if the file is moved.